### ISSUE TITLE
TV input should show when current entity has a TV input (not just when all of them do)

### ISSUE THREAD

**Issue by gdgib** [johanfrick/custom-sonos-card#140]
**Checklist:**

- [x] I updated to the latest version available
- [x] I cleared the cache of my browser

**Release with the issue:** v6

**Last working release (if known):** Pre-v6

**Browser and Operating System:** Edge/Win11 and iOS16.5

**Description of problem:**

When I do not specify `entities` the favorites tab does not show the TV input, even when a TV device is selected (test with both Arc, Beam).  Without `entities`, even a custom source for the TV input does not work.

With `entities` specified, if ALL of the entities have a TV input, then the TV input shows up.  If only some of the entities have a TV input, the TV input is missing.

In short, I think the logic in the JS only shows the TV input if ALL of the entities have a TV input.  Instead, I expect it to show the TV input if the currently selected player has a TV input.

**Javascript errors shown in the web inspector (if applicable):**  I do not know if these are related, but for completeness I checked the console and I see a LOT of these scrolling past.

```
Custom Sonos Card: found no favorites with thumbnails, trying with titles only
custom-sonos-card.js?hacstag=4494168166:392  Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'localeCompare')
    at custom-sonos-card.js?hacstag=4494168166:392:1290
    at Array.sort (<anonymous>)
    at Jt.getAllFavorites (custom-sonos-card.js?hacstag=4494168166:392:1269)
```

**Additional information:**

**Comment by gdgib** [comment]
Turning on "shuffle favorites" shows the TV input.

However, with "shuffle favorites" on, I see three options on the all favorites screen: `TV`, `Line-in`, and a blank unlabelled, but clickable line.  What's more, the three of these keep changing order randomly as I watch.  That is to say, I don't have to reload the page or do anything, they just periodically jump around on the screen.  I'm guessing that's not quite right.  Below you can see two examples, which also highlight the empty line.

![image](https://github.com/johanfrick/custom-sonos-card/assets/801167/706b6b24-4ad8-4315-b4ed-5c97dad1dc64)

![image](https://github.com/johanfrick/custom-sonos-card/assets/801167/f4a24af8-3a13-487f-98a9-3666fad0f79e)

FWIW, I'm a pretty good dev, so I can definitely help debug, I just don't know this code base at all of course.

**Comment by johanfrick** [comment]
Thanks for reporting this bug, I'll take a look.

**Comment by johanfrick** [comment]
I did some initial testing. I can reproduce the shuffle favorites problem, however TV source works as expected for me.
Please share your configuration YAML.

**Comment by johanfrick** [comment]
I've removed shuffleFavorites now. It was buggy, and caused a lot confusion. https://github.com/johanfrick/custom-sonos-card/releases/tag/v6.2

**Comment by gdgib** [comment]
Thank you so much for looking at this so quickly!

Pretty minimal config, shows the random order issue you've removed:

```
type: custom:sonos-card
shuffleFavorites: true
dynamicVolumeSlider: true
```


For the TV inputs...

1. This config shows a TV input on favorites

```
type: custom:sonos-card
shuffleFavorites: true
dynamicVolumeSlider: true
entities
  - media_player.family_room_sonos
```

2. This one shows a TV input, despite there being two possible TV inputs:

```
type: custom:sonos-card
shuffleFavorites: true
dynamicVolumeSlider: true
entities
  - media_player.family_room_sonos
  - media_player.living_room_sonos
```

3. This one shows no TV inputs, despite there being two TV inputs:

```
type: custom:sonos-card
shuffleFavorites: true
dynamicVolumeSlider: true
entities
  - media_player.family_room_sonos
  - media_player.living_room_sonos
  - media_player.kitchen_sonos
```

The family room & kitchen have TVs of course, the kitchen is a sonos one.  I have a few other devices, and as far as I can tell there are two issues:

A. When there's a mix of devices without and without TV inputs, the favorites tab doesn't show any TV inputs.
B. When there are two TV inputs, the favorites tab still just shows "TV" and doesn't distinguish between the two TVs.

I glanced very very briefly at the code, and there seem to be some set membership operations in there.  Any chance the logic is a little off?  If you can't reproduce, that's where I'll start looking if you don't have a better suggestion.

**Comment by johanfrick** [comment]
Not sure I fully understand. You need to first select the player you want to change input/source for. Then go to favorites to select what to source/input to play.
Also, I never see a TV input unless I use customSources, e.g.:
```
customSources: # Main use case is probably to set tv media player to play TV sound
  media_player.tv: # set this to 'all' to show the custom source for all players
    - title: TV
      thumbnail: https://cdn-icons-png.flaticon.com/512/716/716429.png
```

**Comment by gdgib** [comment]
To reproduce:

1. I navigate to groups, and select a player with a TV input (e.g. `media_player.family_room_sonos`)
2. I got to home and confirm that player is listed at the top (e.g. "Family Room Sonos" which is the name of `media_player.family_room_sonos` of course)
3. I got to favorites and see *nothing* except the media browser control in the top right.

This is true regardless of whether there's a custom source or not.  Some more examples I just tried:

Example 1 - No TV input shown because autodiscovery of entities includes sonos players without a TV input, but I expect to see a TV input anyway when selecting players that have a TV input.  It's also really weird that the custom source doesn't show up in this case, but it doesn't regardless of what player I select under "groups".

```
type: custom:sonos-card
dynamicVolumeSlider: true
customSources:
  media_player.family_room_sonos:
    - title: TV
      thumbnail: https://cdn-icons-png.flaticon.com/512/716/716429.png
```

Example 2 - Two TV inputs shown (with different icons) because entities limits me to only sonos players with TV inputs, I guess having two inputs sort of makes sense here, since it has one built in and I'm forcibly adding one.

```
type: custom:sonos-card
dynamicVolumeSlider: true
customSources:
  media_player.family_room_sonos:
    - title: TV
      thumbnail: https://cdn-icons-png.flaticon.com/512/716/716429.png
entities:
  - media_player.family_room_sonos
```

Example 3 - One TV input shown (with no icon) because entities limits me to only sonos players with TV inputs.

```
type: custom:sonos-card
dynamicVolumeSlider: true
entities:
  - media_player.family_room_sonos
```

Bottom line: The only way to get a TV input to appear is to ensure ONLY sonos devices with a TV input and listed in the `entities` config entry.  If ANY sonos device without a TV input is in entities, or I omit `entities` to allow auto-discovery, then the TV inputs don't show regardless of what player I select or custom sources.

**Comment by johanfrick** [comment]
Not sure what is different between your setup and mine. See the video below illustrating that it works as expected on my installation:

https://github.com/johanfrick/custom-sonos-card/assets/26582898/18a686ce-cee1-445e-b5d8-8ca8825ed060

**Comment by gdgib** [comment]
Yeah, that figures. Debugging is fun when you can't reproduce!

Can you give me any pointers on which parts of the code determine what to show in "favorites"? I figure I'll try to do a little debugging and poking around. With a little luck I can at least narrow things down a bit, maybe get you something usable as a unit test at least, or a HAR or something.

I appreciate you confirming it definitely works very differently for us!

**Comment by gdgib** [comment]
Okay, I think this is happening because there's a favorite being returned with an undefined title.  You can see the error I put in the original issue:

```
custom-sonos-card.js?hacstag=4494168166:392  Uncaught (in promise) TypeError: Cannot read properties of undefined (reading 'localeCompare')
    at custom-sonos-card.js?hacstag=4494168166:392:1290
    at Array.sort (<anonymous>)
    at Jt.getAllFavorites (custom-sonos-card.js?hacstag=4494168166:392:1269)
```

When I set a breakpoint at https://github.com/johanfrick/custom-sonos-card/blob/657bcfaca93183b3601ec02caf094cc10782504c/src/sections/media-browser.ts#L153, `a.title` is `undefined`, which of course causes the whole thing to error out hard.  No idea why there's a favorite without a title yet, but at least I can say there's a case here we're not handling.

**Comment by gdgib** [comment]
Okay, it's these lines: https://github.com/johanfrick/custom-sonos-card/blob/657bcfaca93183b3601ec02caf094cc10782504c/src/services/media-browse-service.ts#L66-L68

They assume there's a `source_list` attribute on all sonos media_players, which is not true.  When a player has nothing playing, that attribute disappears, and the resulting `titles` list becomes `[undefined]` when it should be `[]`

**Comment by gdgib** [comment]
Okay, the PR fixes the bug.

That said, I would LOVE to improve the feature.  Right now it just shows "TV" twice on the Favorites list, instead of "TV: Family Room Sonos" and "TV: Living Room Sonos".  But that's a feature, so I'll consider a separate issue once we get this fixed.

**Comment by johanfrick** [comment]
Fixed by #144
